node/core/provisioner/src/disputes/mod.rs

```diff
@@ -0,0 +1,395 @@
+				return result
+			}
+			result.push((session_index, candidate_hash, selected_votes));
+			total_votes_len += votes_len
+		}
+	}
+	result
+}
+
+struct PartitionedDisputes {
+	pub cant_conclude_onchain: Vec<(SessionIndex, CandidateHash)>,
+	pub unknown_onchain: Vec<(SessionIndex, CandidateHash)>,
+	pub can_conclude_onchain: Vec<(SessionIndex, CandidateHash)>,
+	pub concluded_known_onchain: Vec<(SessionIndex, CandidateHash)>, // these will be ignored
+	pub concluded_unknown_onchain: Vec<(SessionIndex, CandidateHash)>,
+}
+
+impl PartitionedDisputes {
+	fn new() -> PartitionedDisputes {
+		PartitionedDisputes {
+			cant_conclude_onchain: Vec::new(),
+			unknown_onchain: Vec::new(),
+			can_conclude_onchain: Vec::new(),
+			concluded_known_onchain: Vec::new(),
+			concluded_unknown_onchain: Vec::new(),
+		}
+	}
+
+	fn into_iter(self) -> impl Iterator<Item = (SessionIndex, CandidateHash)> {
+		self.cant_conclude_onchain
+			.into_iter()
+			.chain(self.unknown_onchain.into_iter())
+			.chain(self.can_conclude_onchain.into_iter())
+			// .chain(self.concluded_known_onchain.into_iter())
+			.chain(self.concluded_unknown_onchain.into_iter())
+	}
+
+	fn disputes_count(&self) -> usize {
+		let PartitionedDisputes {
+			cant_conclude_onchain,
+			unknown_onchain,
+			can_conclude_onchain,
+			concluded_known_onchain: _,
+			concluded_unknown_onchain,
+		} = self;
+		cant_conclude_onchain.len() +
+			unknown_onchain.len() +
+			can_conclude_onchain.len() +
+			concluded_unknown_onchain.len()
+	}
+}
+
+fn partition_recent_disputes(
+	recent: Vec<(SessionIndex, CandidateHash, DisputeStatus)>,
+	onchain: &HashMap<(u32, CandidateHash), DisputeState>,
+) -> PartitionedDisputes {
+	let mut partitioned = PartitionedDisputes::new();
+	let unique_recent = recent
+		.into_iter()
+		.map(|(session_index, candidate_hash, dispute_state)| {
+			((session_index, candidate_hash), dispute_state)
+		})
+		.collect::<HashMap<_, _>>();
+
+	let (active, concluded): (
+		Vec<(SessionIndex, CandidateHash, DisputeStatus)>,
+		Vec<(SessionIndex, CandidateHash, DisputeStatus)>,
+	) = unique_recent
+		.into_iter()
+		.map(|((session_index, candidate_hash), dispute_state)| {
+			(session_index, candidate_hash, dispute_state)
+		})
+		.partition(|(_, _, status)| *status == DisputeStatus::Active);
+
+	for (session_index, candidate_hash, _) in active {
+		match onchain.get(&(session_index, candidate_hash)) {
+			Some(d) => {
+				// TODO: handle double voting here!!!
+				if d.validators_for.count_ones() + d.validators_against.count_ones() >
+					supermajority_threshold(d.validators_for.len())
+				{
+					partitioned.can_conclude_onchain.push((session_index, candidate_hash));
+				} else {
+					partitioned.cant_conclude_onchain.push((session_index, candidate_hash));
+				}
+			},
+			None => partitioned.unknown_onchain.push((session_index, candidate_hash)),
+		};
+	}
+
+	// TODO: Prioritise the votes in each partition. Is it still necessary?
```

Comment:
We can do even more here:

> Provisioner can send to the runtime just enough votes so that the dispute can conclude. This way votes for more disputes can be pushed to the runtime.

> Provisioner can prioritise disputes with more votes because they have got higher chance for fast conclusion.

But is this necessary?

Reply:
We should not over optimize.

* sending too few votes will not reward just ever so slightly late dispute statements or include them, this means work without reward - from a sec point that's fine, but we should try to maximize the number of votes included
* the biggest issue with disputes is being time bound it's of utmost concern to conclude older disputes. That said, if there are multiple disputes for candidates of the same relay chain block, it might make sense to do so in that scope. I am not sure how async backing interferes with that, so I'd not over optimize here until that's clear.